src/sinks/splunk_hec/common/service.rs

```diff
@@ -74,13 +150,265 @@ pub struct HttpRequestBuilder {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            ..HecClientAcknowledgementsConfig::default()
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let request = get_hec_request();
+        let response = service.call(request).await.unwrap();
+        assert_eq!(EventStatus::Delivered, response.event_status)
+    }
+
+    #[tokio::test]
+    async fn acknowledgements_enabled_on_server_retry_limit_exceeded() {
+        let ack_response = |req: &Request| {
+            let req = serde_json::from_slice::<HecAckStatusRequest>(req.body.as_slice()).unwrap();
+            ResponseTemplate::new(200).set_body_json(HecAckStatusResponse {
+                acks: req
+                    .acks
+                    .into_iter()
+                    .map(|ack_id| (ack_id, false))
+                    .collect::<HashMap<_, _>>(),
+            })
+        };
+        let mock_server = get_hec_mock_server(true, ack_response).await;
+
+        let acknowledgements_config = HecClientAcknowledgementsConfig {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            retry_limit: NonZeroU8::new(1).unwrap(),
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let request = get_hec_request();
+        let response = service.call(request).await.unwrap();
+        assert_eq!(EventStatus::Failed, response.event_status)
+    }
+
+    #[tokio::test]
+    async fn acknowledgements_server_changed_ack_response_format() {
+        let ack_response = |_: &Request| {
+            ResponseTemplate::new(200)
+                .set_body_json(serde_json::json!(r#"{ "new": "a new response body" }"#))
+        };
+        let mock_server = get_hec_mock_server(true, ack_response).await;
+
+        let acknowledgements_config = HecClientAcknowledgementsConfig {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            retry_limit: NonZeroU8::new(3).unwrap(),
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let request = get_hec_request();
+        let response = service.call(request).await.unwrap();
+        assert_eq!(EventStatus::Delivered, response.event_status)
+    }
+
+    #[tokio::test]
+    async fn acknowledgements_enabled_on_server_ack_endpoint_failing() {
+        let ack_response = |_: &Request| ResponseTemplate::new(503);
+        let mock_server = get_hec_mock_server(true, ack_response).await;
+
+        let acknowledgements_config = HecClientAcknowledgementsConfig {
+            query_interval: NonZeroU8::new(1).unwrap(),
+            retry_limit: NonZeroU8::new(3).unwrap(),
+        };
+        let mut service = get_hec_service(mock_server.uri(), acknowledgements_config);
+
+        let request = get_hec_request();
+        let response = service.call(request).await.unwrap();
+        assert_eq!(EventStatus::Failed, response.event_status)
```

Comment:
Should this be `Errored`, as the server didn't explicitly reject the events?

Reply:
I see. I've had a hard time with the error handling, so it's somewhat messy. In this case, an `ack_id` would have been queried `retry_limit` number of times and never been successfully acked. It would then be dropped and result in `EventStatus::Failed` since we don't keep track of whether it hasn't been acked due to `ack_id: false` status or because the endpoint is erroring. 

Supporting `EventStatus::Errored` seems more accurate but I'm curious if you have suggestions for the implementation.

Reply:
I think `Errored` is more accurate as well as we don't know why the events were never acked.

Reply:
Updated in e81824d